src/model_theory/definability.lean

```diff
@@ -82,95 +173,195 @@ lemma is_definable.compl {s : set (α → M)} (hf : L.is_definable s) :
+    ext x,
+    rw function.comp_apply,
+    refine congr rfl _,
+    ext,
+    rw fin.coe_cast_add },
+end
+
+/-- Shows that definability is closed under general projections. -/
+lemma definable.image_comp {s : set (β → M)} (h : L.definable A s)
+  (f : α → β) :
+  L.definable A ((λ g : β → M, g ∘ f) '' s) :=
+begin
+  classical,
+  have h := h.image_comp_equiv (equiv.trans (equiv.sum_congr (_root_.equiv.refl _)
+    (fintype.equiv_fin _).symm) (equiv.set.sum_compl (range f))),
+  have h := (h.image_comp_sum_inl_fin _).preimage_comp (range_splitting f),
+  have h' : L.definable A ({ x : α → M |
+    ∀ a, x a = x (range_splitting f (range_factorization f a))}),
+  { have h' : ∀ a, L.definable A {x : α → M | x a =
+      x (range_splitting f (range_factorization f a))},
+    { refine λ a, ⟨⟨(var a).equal (var (range_splitting f (range_factorization f a))), ext _⟩⟩,
+      simp, },
+    refine (congr rfl (ext _)).mp (definable_finset_bInter h' finset.univ),
+    simp },
+  refine (congr rfl (ext (λ x, _))).mp (h.inter h'),
+  simp only [equiv.coe_trans, mem_inter_eq, mem_preimage, mem_image,
+    exists_exists_and_eq_and, mem_set_of_eq],
+  split,
+  { rintro ⟨⟨y, ys, hy⟩, hx⟩,
+    refine ⟨y, ys, _⟩,
+    ext a,
+    rw [hx a, ← function.comp_apply x, ← hy],
+    simp, },
+  { rintro ⟨y, ys, rfl⟩,
+    refine ⟨⟨y, ys, _⟩, λ a, _⟩,
+    { ext,
+      simp [set.apply_range_splitting f] },
+    { rw [function.comp_apply, function.comp_apply, apply_range_splitting f,
+        range_factorization_coe], }}
+end
+
+variables (L) {M} (A)
+
+/-- A 1-dimensional version of `definable`, for `set M`. -/
+def definable₁ (s : set M) : Prop := L.definable A { x : fin 1 → M | x 0 ∈ s }
+
+/-- A 2-dimensional version of `definable`, for `set (M × M)`. -/
+def definable₂ (s : set (M × M)) : Prop := L.definable A { x : fin 2 → M | (x 0, x 1) ∈ s }
+
+variables (α)
 
 /-- Definable sets are subsets of finite Cartesian products of a structure such that membership is
   given by a first-order formula. -/
-def definable_set := subtype (λ s : set (α → M), is_definable L s)
+def definable_set := subtype (λ s : set (α → M), L.definable A s)
 
 namespace definable_set
-variables {M} {α}
+variables {A} {α}
 
-instance : has_top (L.definable_set M α) := ⟨⟨⊤, is_definable_univ⟩⟩
+instance : has_top (L.definable_set A α) := ⟨⟨⊤, definable_univ⟩⟩
 
-instance : has_bot (L.definable_set M α) := ⟨⟨⊥, is_definable_empty⟩⟩
+instance : has_bot (L.definable_set A α) := ⟨⟨⊥, definable_empty⟩⟩
 
-instance : inhabited (L.definable_set M α) := ⟨⊥⟩
+instance : inhabited (L.definable_set A α) := ⟨⊥⟩
 
-instance : set_like (L.definable_set M α) (α → M) :=
+instance : set_like (L.definable_set A α) (α → M) :=
 { coe := subtype.val,
   coe_injective' := subtype.val_injective }
 
 @[simp]
-lemma mem_top {x : α → M} : x ∈ (⊤ : L.definable_set M α) := set.mem_univ x
+lemma mem_top {x : α → M} : x ∈ (⊤ : L.definable_set A α) := mem_univ x
```

Comment:
I encourage you to split your PRs into "new material" and "cosmetic changes" in the future.
This part of the diff is completely uncontroversial, and I would be happy to merge it straight away.
As a result, the diff on "new material" would be smaller, and easier to review in detail.

Reply:
I will definitely take this advice in principle, and will see if it applies to my current PRs.

However, in this precise case, this line contains two changes. One is removing the `set.`, which is indeed cosmetic, while the other is the change from `M` to a parameter set `A`, which is taking into account the new features of this PR.

Reply:
Still, I can try to split this PR into some easier pieces.

Reply:
In fact, I have done so. I have spun off two sub-PRs, one on the required `Lhom` operations without any definability, and one with the definition and nomenclature changes.